crypto/aria/aria.c

```diff
@@ -1,454 +1,1289 @@
-        break;
-    }
-
-    FO(w3, w0, ck1);	xor128(w1, w3, kr);
-    FE(w3, w1, ck2);	xor128(w2, w3, w0);
-    FO(kr, w2, ck3);	xor128(w3, kr, w1);
-
-    rot19r(key->rd_key[ 0], w0, w1);
-    rot19r(key->rd_key[ 1], w1, w2);
-    rot19r(key->rd_key[ 2], w2, w3);
-    rot19r(key->rd_key[ 3], w3, w0);
-
-    rot31r(key->rd_key[ 4], w0, w1);
-    rot31r(key->rd_key[ 5], w1, w2);
-    rot31r(key->rd_key[ 6], w2, w3);
-    rot31r(key->rd_key[ 7], w3, w0);
-
-    rot61l(key->rd_key[ 8], w0, w1);
-    rot61l(key->rd_key[ 9], w1, w2);
-    rot61l(key->rd_key[10], w2, w3);
-    rot61l(key->rd_key[11], w3, w0);
-
-    rot31l(key->rd_key[12], w0, w1);
-    if (key->rounds > 12) {
-        rot31l(key->rd_key[13], w1, w2);
-        rot31l(key->rd_key[14], w2, w3);
-
-        if (key->rounds > 14) {
-            rot31l(key->rd_key[15], w3, w0);
-            rot19l(key->rd_key[16], w0, w1);
-        }
-    }
-    return 0;
-}
-
-/*
- * Expand the cipher key into the decryption key schedule.
- */
-int aria_set_decrypt_key(const unsigned char *userKey, const int bits,
-                         ARIA_KEY *key)
-{
-    ARIA_KEY ek;
-    const int r = aria_set_encrypt_key(userKey, bits, &ek);
-    unsigned int i, rounds = ek.rounds;
-
-    if (r == 0) {
-        key->rounds = rounds;
-        memcpy(key->rd_key[0], ek.rd_key[rounds], sizeof(key->rd_key[0]));
-        for (i = 1; i < rounds; i++)
-            a(key->rd_key[i], ek.rd_key[rounds - i]);
-        memcpy(key->rd_key[rounds], ek.rd_key[0], sizeof(key->rd_key[rounds]));
-    }
-    return r;
-}
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* 
+ * Copyright (C) 2017 National Security Research Institute. All Rights Reserved.
+ *
+ * Information for ARIA
+ *     http://210.104.33.10/ARIA/index-e.html (English)
+ *     http://seed.kisa.or.kr/ (Korean)
+ *
+ * Public domain version is distributed above.
+ */
+
+/* ====================================================================
+ * Copyright (c) 2017 Oracle and/or its affiliates.  All rights reserved.
+ */
+
+#include <openssl/e_os2.h>
+#include "internal/aria.h"
+
+#include <assert.h>
+#include <string.h>
+#include <stddef.h>
+
+#ifndef OPENSSL_SMALL_FOOTPRINT
+
+/* Begin macro */
+#ifndef __has_builtin
+# define __has_builtin(x) 0
+#endif
+
+#if !defined(B_ENDIAN) && !defined(L_ENDIAN)
+# if defined(__BYTE_ORDER__) && defined(__ORDER_BIG_ENDIAN__) && \
+    (__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__)
+#  define B_ENDIAN
+# elif defined(__BIG_ENDIAN__) || \
+    defined(_M_PPC) || defined(_M_MPPC) || \
+    defined(__ARMEB__) || defined(__THUMBEB__) || defined(__AARCH64EB__) || \
+    defined(_MIPSEB) || defined(__MIPSEB) || defined(__MIPSEB__)

```

Comment:
Trouble with this approach is that it makes you track down every big-endian system. And above list is far from complete. Since it's big-endian algorithm, it should be vice-versa, i.e. it should work without B_ENDIAN or L_ENDIAN, and *then* one can "opt-in" known to be little-endian ones. This way you ensure that even if list is incomplete, it still produces correct result.

With this in mind I would actually suggest (rather insist) that this initial submission is free from *any* platform-specific #if-spaghetti. And I literally mean any, not even bswap thing. If proven critical, the #if-spaghetti can be added as follow-up commit.